README.md
# pyfintech

CHANGELOG.txt
Change Log
-----------
-----------

0.1 (25/02/2023)
------------------
-First Released of library


ipo_details.py
import datetime
import yfinance as yf

def ipo_date(ticker):
    # Use the Ticker class from the yfinance library to get the historical data for the stock
    stock = yf.Ticker(ticker.upper())
    # Get the maximum period of historical data available for the stock
    stock_history = stock.history(period="max")
    # Get the earliest available data, which is the first row of the stock_history data frame
    ipo_day = stock_history.head(1)
    # Get the year, month, and day of the IPO date
    year = ipo_day.index[0].year
    month = ipo_day.index[0].month
    day = ipo_day.index[0].day
    # Print the IPO date in the format "Company ticker '{ticker}' has Lauch their IPO on {day}-{month}-{year}"
    print(f"Company ticker '{ticker.upper()}' has Lauch their IPO on {day}-{month}-{year}")


'''The IPO tenure refers to the length of time that a company has been publicly traded since 
its initial public offering (IPO). The IPO tenure begins on the date of the company's IPO and
 continues until the present day.

The IPO tenure can be an important factor for investors as it provides information about the company's financial performance 
and stability since it went public. 

Companies that have been publicly traded for a long time are generally seen as being more established 
and credible than companies that have only recently gone public. 

The IPO tenure can also provide insight into the company's ability to attract 
and retain investors, as well as its ability to navigate the challenges of being a publicly traded company.

It's important to note that the IPO tenure is only one of many factors that investors should consider when evaluating a publicly traded company.
Other factors, such as the company's financial performance, management team, and industry conditions, 
are also important to consider when making investment decisions.'''

def ipo_tenture(ticker):

    # Use the Ticker class from the yfinance library to get the historical data for the stock
    stock = yf.Ticker(ticker.upper())
    # Get the maximum period of historical data available for the stock
    stock_history = stock.history(period="max")
    # Get the earliest available data, which is the first row of the stock_history data frame
    ipo_day = stock_history.head(1)
    # Get the year, month, and day of the IPO date
    year = ipo_day.index[0].year
    month = ipo_day.index[0].month
    day = ipo_day.index[0].day
    
    # Specify the date you want to calculate the difference from
    target_date = datetime.datetime(year, month, day)

    # Calculate the difference between the target date and today
    today = datetime.datetime.now()
    difference = today - target_date 

    # Print the number of days between today and the target date
    print(f"Number of days between today and {target_date.strftime('%Y-%m-%d')}: {difference.days} days")



def ipo_price(ticker):
    # Use the Ticker class from the yfinance library to get the historical data for the stock
    stock = yf.Ticker(ticker.upper())
    # Get the maximum period of historical data available for the stock
    stock_history = stock.history(period="max")
    # Get the earliest available data, which is the first row of the stock_history data frame
    ipo_inital = stock_history.Open[0]
    final_price = stock_history.Close[-1]

    # Print the number of days between today and the target date
    print(f"Initial price on launch {ipo_inital} and current price {final_price}")


price_status.py
# import yfinance as yf
# import numpy as np
# import pyttsx3
# import pandas as pd

# class TextToSpeech:
#     def __init__(self, text, rate=140, volume=1.0):
#         self.text = text
#         self.rate = rate
#         self.volume = volume
#         self.engine = pyttsx3.init()

#     def speak(self):
#         self.engine.setProperty('rate', self.rate)
#         self.engine.setProperty('volume', self.volume)
#         self.engine.say(self.text)
#         self.engine.runAndWait()
#         self.engine.stop()

#     def save_to_file(self, file_name):
#         self.engine.save_to_file(self.text, file_name)
#         self.engine.runAndWait()

# def analyze_stocks(stock_list):
#     # Mine data from the Yahoo Finance API and store it as a list of dataframes
#     stock_insight = [yf.Ticker(i).history(period="365d") for i in stock_list]

#     # Define a function to perform 7-day analysis on a stock dataframe
#     def seven_days(df):
#         open_minimum = df.tail(7).describe()[3:4]['Open'][0]
#         close_minimum = df.tail(7).describe()[3:4]['Close'][0]
#         open_maximum = df.tail(7).describe()[7:8]['Open'][0]
#         close_maximum = df.tail(7).describe()[7:8]['Close'][0]
#         open_mean = df.tail(7)["Open"].mean()
#         close_mean = df.tail(7)["Close"].mean()
#         open_variance = np.var(df.tail(7)["Open"])
#         close_variance = np.var(df.tail(7)["Close"])
#         open_stddev = np.std(df.tail(7)["Open"])
#         close_stddev = np.std(df.tail(7)["Close"])
#         return (open_minimum, close_minimum, open_maximum, close_maximum, open_mean, close_mean, open_variance, close_variance, open_stddev, close_stddev)

#     # Use the map function to apply the seven_days function to all stock dataframes in the stock_insight list
#     seven_day_analysis = list(map(seven_days, stock_insight))

#     # Use the zip function to combine the results with the stock list and create a list of tuples
#     result = list(zip(stock_list, seven_day_analysis))

#     for stock, analysis in result:
#         open_min, close_min, open_max, close_max, open_mean, close_mean, open_variance, close_variance, open_stddev, close_stddev = analysis
#         print(f"Stock: {stock}")
#         print(f"7-day Open mean:\t\t{round(open_mean,2)}")
#         print(f"7-day Close mean:\t\t{round(close_mean,2)}")
#         print(f"7-day Open variance:\t{round(open_variance,2)}")
#         print(f"7-day Close variance:\t{round(close_variance,2)}")
#         print(f"7-day Open standard deviation:\t{round(open_stddev,2)}")
#         print(f"7-day Close standard deviation:\t{round(close_stddev,2)}")
#         print(f"7-day Open maximum:\t\t{round(open_max,2)}")
#         print(f"7-day Open minimum:\t\t{round(open_min,2)}")
#         print(f"7-day Close maximum:\t\t{round(close_max,2)}")
#         print(f"7-day Close minimum:\t\t{round(close_min,2)}")
#         print()
#         tts = TextToSpeech(f'{stock} has a mean Open price of {round(open_mean,2)} dollars in the last 7 days, with a minimum Open price of {round(open_min,2)} dollars, a maximum Open price of {round(open_max,2)} dollars, and a standard deviation of {round(open_stddev,2)} dollars')
#         tts.speak()
#         tts.save_to_file(f"{stock}.mp3")

import yfinance as yf
import numpy as np
import pyttsx3
import pandas as pd


class TextToSpeech:
    def __init__(self, text, rate=140, volume=1.0):
        self.text = text
        self.rate = rate
        self.volume = volume
        self.engine = pyttsx3.init()

    def speak(self):
        self.engine.setProperty('rate', self.rate)
        self.engine.setProperty('volume', self.volume)
        self.engine.say(self.text)
        self.engine.runAndWait()
        self.engine.stop()

    def save_to_file(self, file_name):
        self.engine.save_to_file(self.text, file_name)
        self.engine.runAndWait()


class StockAnalysis:
    def __init__(self, stock_list):
        self.stock_list = stock_list
        self.stock_insight = [yf.Ticker(i).history(
            period="365d") for i in self.stock_list]

    def analyze(self):
        # Define a function to perform 7-day analysis on a stock dataframe
        def seven_days(df):
            open_minimum = df.tail(7).describe()[3:4]['Open'][0]
            close_minimum = df.tail(7).describe()[3:4]['Close'][0]
            open_maximum = df.tail(7).describe()[7:8]['Open'][0]
            close_maximum = df.tail(7).describe()[7:8]['Close'][0]
            open_mean = df.tail(7)["Open"].mean()
            close_mean = df.tail(7)["Close"].mean()
            open_variance = np.var(df.tail(7)["Open"])
            close_variance = np.var(df.tail(7)["Close"])
            open_stddev = np.std(df.tail(7)["Open"])
            close_stddev = np.std(df.tail(7)["Close"])
            return (open_minimum, close_minimum, open_maximum, close_maximum, open_mean, close_mean, open_variance, close_variance, open_stddev, close_stddev)

        # Use the map function to apply the seven_days function to all stock dataframes in the stock_insight list
        seven_day_analysis = list(map(seven_days, self.stock_insight))

        # Use the zip function to combine the results with the stock list and create a list of tuples
        result = list(zip(self.stock_list, seven_day_analysis))

        for stock, analysis in result:
            open_min, close_min, open_max, close_max, open_mean, close_mean, open_variance, close_variance, open_stddev, close_stddev = analysis
            print(f"Stock: {stock}")
            print(f"7-day Open mean:\t\t{round(open_mean,2)} dollars")
            print(f"7-day Close mean:\t\t{round(close_mean,2)} dollars")
            print(f"7-day Open variance:\t\t{round(open_variance,2)} dollars")
            print(
                f"7-day Close variance:\t\t{round(close_variance,2)} dollars")
            print(
                f"7-day Open standard deviation:\t{round(open_stddev,2)} dollars")
            print(
                f"7-day Close standard deviation:\t{round(close_stddev,2)} dollars")
            print(
                f"7-day Open maximum:\t\t{round(open_max,2)} dollars ({self.stock_insight[0].index[-1].strftime('%Y-%m-%d')})")
            print(
                f"7-day Open minimum:\t\t{round(open_min,2)} dollars ({self.stock_insight[0].index[-7].strftime('%Y-%m-%d')})")
            print(
                f"7-day Close maximum:\t\t{round(close_max,2)} dollars ({self.stock_insight[0].index[-1].strftime('%Y-%m-%d')})")
            print(
                f"7-day Close minimum:\t\t{round(close_min,2)} dollars ({self.stock_insight[0].index[-7].strftime('%Y-%m-%d')})")
            print()
            tts = TextToSpeech(
                f'Stock {stock} has a mean Open price of {round(open_mean,2)} dollars in the last 7 days. The minimum Open price was {round(open_min,2)} dollars on {self.stock_insight[0].index[-7].strftime("%Y-%m-%d")}. The maximum Open price was {round(open_max,2)} dollars on {self.stock_insight[0].index[-1].strftime("%Y-%m-%d")}. The standard deviation of the Open price was {round(open_stddev,2)} dollars.')
            tts.speak()
            tts.save_to_file(f"{stock}.mp3")

# analysis = StockAnalysis(['AAPL', 'GOOG', 'MSFT', 'AMZN'])
# analysis.analyze()


ThresholdClassifier.py
from typing import List
import matplotlib.pyplot as plt
import numpy as np
# class ThresholdClassifier:
#     def __init__(self, threshold: float):
#         self.threshold = threshold
#     def classify(self, data: List[float]):
#         self.mean = sum(data) / len(data)
#         self.threshold = self.mean
#         return [1 if x >= self.threshold else 0 for x in data]  
# classifier = ThresholdClassifier(0.5)

class ThresholdClassifier:
    def __init__(self, threshold: float):
        self.threshold = threshold
        
    def classify(self, data: List[float]):
        self.mean = sum(data) / len(data)
        self.threshold = self.mean
        classifications = []
        for value in data:
            if value > self.threshold:
                classifications.append(1)
            else:
                classifications.append(0)
        return classifications
    
classifier = ThresholdClassifier(0.5)

class ThresholdPlot:
    def __init__(self, data, classification):
        self.data = data
        self.classification = classification
        
    def plot(self):
        plt.style.use('dark_background')
        plt.figure(figsize=(15,6))
        plt.xlabel("Time", fontsize=14)
        plt.ylabel("Price (£)", color="tab:red", fontsize=14)
        plt.plot(self.data.index, self.data, color="tab:red")
        plt.twinx()
        plt.ylabel("Classification", color="tab:blue", fontsize=14)
        plt.plot(self.data.index, self.classification, color="tab:blue")
        plt.axhline(y=0.5, color="gray", linestyle="--")
        plt.title("Binary Classifications", fontsize=20, color="White")
        plt.show()

        colors = []
        for value in self.classification:
            if value > 0.5:
                colors.append("tab:green")
            else:
                colors.append("tab:red")

        plt.style.use('dark_background')
        plt.figure(figsize=(15,7))
        plt.xlabel("Time", fontsize=14)
        plt.ylabel("Price (£)", color="tab:red", fontsize=14)
        plt.plot(self.data.index, self.data, color="tab:red")
        plt.grid(linestyle="--", color="gray", alpha=0.7)
        plt.tick_params(axis="both", labelsize=12)
        plt.twinx()
        plt.ylabel("Classification", color="tab:blue", fontsize=14)
        plt.scatter(self.data.index, self.classification, color=colors)
        plt.tick_params(axis="both", labelsize=12)
        plt.title("Binary Classifications", fontsize=18, color="White")
        plt.show()





VolatilityModel.py
import yfinance as yf
import numpy as np
import pyttsx3
import pandas as pd
import matplotlib.pyplot as plt


class TextToSpeech:
    def __init__(self, text, rate=140, volume=1.0):
        self.text = text
        self.rate = rate
        self.volume = volume
        self.engine = pyttsx3.init()

    def speak(self):
        self.engine.setProperty('rate', self.rate)
        self.engine.setProperty('volume', self.volume)
        self.engine.say(self.text)
        self.engine.runAndWait()
        self.engine.stop()

    def save_to_file(self, file_name):
        self.engine.save_to_file(self.text, file_name)
        self.engine.runAndWait()


# def stock_volatility(stock_ticker, days):
#     stock_data = yf.Ticker(stock_ticker).history(period=f"{days}d")
#     daily_returns = stock_data["Close"].pct_change().dropna()
#     standard_deviation = np.std(daily_returns)

#     if standard_deviation < 0.05:
#         result = f"The standard deviation of {stock_ticker}'s daily returns is {standard_deviation:.4f}. This stock can be considered relatively safe to trade."
#     elif standard_deviation >= 0.05 and standard_deviation < 0.1:
#         result = f"The standard deviation of {stock_ticker}'s daily returns is {standard_deviation:.4f}. This stock is moderately safe to trade."
#     else:
#         result = f"The standard deviation of {stock_ticker}'s daily returns is {standard_deviation:.4f}. This stock is considered risky to trade."

#     # Plot the daily returns
#     plt.plot(daily_returns, label=f"{stock_ticker} Daily Returns")
#     # Plot the standard deviation
#     plt.axhline(standard_deviation, color='red',
#                 label=f"Standard Deviation ({standard_deviation:.4f})")
#     # Add title, labels, and legend
#     plt.title(f"{stock_ticker} Daily Returns and Standard Deviation")
#     plt.xlabel("Time")
#     plt.ylabel("Returns")
#     plt.legend()
#     # Show the plot
#     plt.show()

#     print(result)
#     tts = pyttsx3.init()
#     tts.setProperty('rate', 160)
#     tts.say(result)
#     tts.runAndWait()


# def volatility_test(stock_list, days):

    # for stock in stock_list:
    #     print(f"\nStock: {stock}")
    #     stock_volatility(stock, days)


# stock_list = ['AAPL', 'GOOG', 'MSFT', 'AMZN']
# volatility_test(stock_list, "1y")


def stock_volatility(close_prices):
    '''The number of days to consider for a volatility test in stock market depends on various factors such as the type of security, 
    the time frame of the analysis, and the level of detail you want to achieve. 
    However, a common practice is to consider at least 60 to 252 trading days (i.e., roughly 3 to 12 months) to capture the short-term and long-term volatility patterns.
    This time frame allows for a good representation of the market's ups and downs and provides a reliable measure of the stock's historical volatility.

    It's important to note that the stock market is inherently volatile and it's common to see fluctuations in prices on a daily basis. 
    By using a longer time frame, you can get a better understanding of the underlying trends and patterns in the stock's price movements. 
    However, it's also important to keep in mind that past performance is not indicative of future results,
    and the stock's volatility can change over time due to various factors such as economic conditions, company-specific events, and market sentiment.'''

    '''Day trading, it's common to consider a shorter time frame when measuring standard deviation, such as the last 20 to 50 trading days. 
    This time frame provides a good representation of the stock's short-term volatility, 
    which is important in day trading where the goal is to take advantage of price movements over a single day or a few days.

    However, it's important to keep in mind that the stock market is inherently volatile and it's common to see fluctuations in prices on a daily basis. 
    By using a shorter time frame, you may miss important trends and patterns in the stock's price movements. 
    It's a good idea to consider a longer time frame as well to get a better understanding of the stock's overall volatility.

    Ultimately, the number of days to consider for measuring standard deviation depends on your trading strategy, the type of security, 
    and the level of detail you want to achieve. It's a good idea to experiment with different time frames and see which one works best for your particular situation.'''
    '''For long-term holders, it's common to consider a longer time frame when measuring standard deviation, such as the last 5 to 10 years. 
    
    This time frame provides a good representation of the stock's long-term volatility and allows for a more comprehensive analysis of the stock's price movements over a longer period of time.

    A lower standard deviation indicates that the stock has been less volatile over the long-term, while a higher standard deviation indicates greater volatility. 
    Long-term holders typically look for stocks with lower volatility as they are less likely to experience large price swings over a longer period of time.

    However, it's important to keep in mind that the stock market is inherently volatile and it's common to see fluctuations in prices over a long period of time. 
    By using a longer time frame, you may miss important short-term trends and patterns in the stock's price movements. 
    It's a good idea to consider a shorter time frame as well to get a better understanding of the stock's overall volatility.

    Ultimately, the number of days to consider for measuring standard deviation depends on your investment strategy, the type of security, and the level of detail you want to achieve. 
    It's a good idea to experiment with different time frames and see which one works best for your particular situation.'''

    daily_returns = close_prices.pct_change().dropna()
    standard_deviation = np.std(daily_returns)

    if standard_deviation < 0.05:
        result = f"The standard deviation of the stock's daily returns is {standard_deviation:.4f}. This stock can be considered relatively safe to trade."
    elif standard_deviation >= 0.05 and standard_deviation < 0.1:
        result = f"The standard deviation of the stock's daily returns is {standard_deviation:.4f}. This stock is moderately safe to trade."
    else:
        result = f"The standard deviation of the stock's daily returns is {standard_deviation:.4f}. This stock is considered risky to trade."

    # Plot the daily returns
    plt.figure(figsize=(15, 6))
    plt.plot(daily_returns, label="Stock Daily Returns")
    # Plot the standard deviation
    plt.axhline(standard_deviation, color='red',
                label=f"Standard Deviation ({standard_deviation:.4f})")
    # Add title, labels, and legend
    plt.title("Stock Daily Returns and Standard Deviation")
    plt.xlabel("Time")
    plt.ylabel("Returns")
    plt.legend()
    # Show the plot
    plt.show()

    print(result)
    tts = pyttsx3.init()
    tts.setProperty('rate', 160)
    tts.say(result)
    tts.runAndWait()
